common/windows/delphi/general/Upload_Settings.pas

```diff
@@ -169,7 +170,7 @@ function URLPath_PackageDownload(const PackageID, BCP47: string; IsUpdate: Boole
 begin
   if IsUpdate then IsUpdateInt := 1 else IsUpdateInt := 0;
 
-  Result := Format(URLPath_PackageDownload_Format, [PackageID, CKeymanVersionInfo.Tier, BCP47, IsUpdateInt]);
+  Result := Format(URLPath_PackageDownload_Format, [TIdURI.URLEncode(PackageID), CKeymanVersionInfo.Tier, BCP47, IsUpdateInt]);
```

Comment:
I would have gone with 
```
Result := TIdURI.URLEncode(Format(URLPath_PackageDownload_Format, [PackageID, CKeymanVersionInfo.Tier, BCP47, IsUpdateInt]));
```
However, after Marc commented on the URI encoding library issues and looking a bit further myself I decided it was safer to just tackle the segment of the URI we know to be an issue and not introduce possible bugs due to incorrect encoding of the rest of the URI address.

Reply:
Note, it's logically impossible to uri encode an entire URI. Imagine PackageID has a `/` in it. How would `TIdURI.URLEncode` know to encode that `/` and not the ones that are in `URLPath_PackageDownload_Format`? We have to encode the components before injecting them into the URL.

For complete correctness, we should encode each string parameter in the format string:

```
 Result := Format(URLPath_PackageDownload_Format, [TIdURI.URLEncode(PackageID), TIdURI.URLEncode(CKeymanVersionInfo.Tier), TIdURI.URLEncode(BCP47), IsUpdateInt]);
```

Now, to make your life even more sad, read this SO: https://stackoverflow.com/questions/776302/standard-url-encode-function

If you can figure out the right answer from that, you're doing better than I am... and I am referenced in some of the answers. It's a mess.

Currently, we try to use https://github.com/keymanapp/keyman/blob/6eefb3a771ea417fadda127b5ff23d099e761510/common/windows/delphi/general/utilhttp.pas#L46-L98 everywhere, and we avoid TIdURI entirely. That function has been updated just over a year ago, so I am reasonably confident that it will cover all the bases. (Yeah nah unit tests would be nice?!)